pyangbind/lib/serialise.py

```diff
@@ -164,54 +160,271 @@ def default(self, obj, mode="default"):
             else:
-                if obj:
-                    return True
-                else:
-                    return False
+                return bool(obj)
         elif map_val in ["pyangbind.lib.yangtypes.TypedList"]:
-            return [self.default(i) for i in obj]
-        elif map_val in ["int"]:
-            int_size = getattr(obj, "_restricted_int_size", None)
-            if mode == "ietf" and int_size == 64:
-                return six.text_type(obj)
-            return int(obj)
-        elif map_val in ["long"]:
+            return self.yangt_typed_list(obj)
+        elif map_val in ["int", "long"]:
             int_size = getattr(obj, "_restricted_int_size", None)
-            if mode == "ietf" and int_size == 64:
-                return six.text_type(obj)
-            elif six.PY3:
-                return int(obj)
-            else:
-                return long(obj)
+            return self.yangt_long(obj) if int_size == 64 else self.yangt_int(obj)
         elif map_val in ["container"]:
-            return self._preprocess_element(obj.get(), mode=mode)
+            return self.preprocess_element(obj.get())
         elif map_val in ["decimal.Decimal"]:
-            return six.text_type(obj) if mode == "ietf" else float(obj)
+            return self.yangt_decimal(obj)
+
+    def yangt_int(self, obj):
+        # for values that are 32-bits and under..
+        return int(obj)
+
+    def yangt_long(self, obj):
+        # don't need to sepcial-case for PY3 because we assign `long = int` in the header
+        return long(obj)
+
+    def yangt_identityref(self, obj):
+        raise UnmappedItem
+
+    def yangt_decimal(self, obj):
+        # doesn't seem to need a distinction for non-64-bit decimal?
+        return float(obj)
+
+    def yangt_empty(self, obj):
+        return bool(obj)
+
+    def yangt_typed_list(self, obj):
+        return [self.default(i) for i in obj]
+
+
+class IETFYangDataSerialiser(YangDataSerialiser):
+    """
+    IETF data serialiser overrides some of the data type formats only
+    """
+
+    def yangt_long(self, obj):
+        return six.text_type(obj)
+
+    def yangt_identityref(self, obj):
+        try:
+            emod = obj._enumeration_dict[obj]["@module"]
+            if emod != obj._defining_module:
+                return "%s:%s" % (obj._enumeration_dict[obj]["@module"], obj)
+        except KeyError:
+            pass
+        return six.text_type(obj)
+
+    def yangt_decimal(self, obj):
+        return six.text_type(obj)
+
+    def yangt_empty(self, obj):
+        return [None] if obj else False
+
+
+class XmlYangDataSerialiser(IETFYangDataSerialiser):
+    """
+    XML can have an empty tag, and a TypedList must be treated specially, here we mark it with a deque
+    """
+
+    def yangt_typed_list(self, obj):
+        return deque([self.default(i) for i in obj])
+
+    def yangt_empty(self, obj):
+        return None
+
+
+class pybindJSONEncoderBase(json.JSONEncoder):
```

Comment:
Here and below, it'd be useful to clarify for a reader what these different classes are for.